Add first tutorial lessons (#484)
crystal-lang/crystal-book: docs/tutorials/01_basics/02_variables.md

# Variables

In order to store a value and re-use it later, it can be assigned to a variable.

For example, if you want to say `Hello Penny!` three times, you don't need to repeat the same string multiple times.
Instead, you can assign it to a variable and re-use it:

```{.crystal .crystal-play}
message = "Hello Penny!"

puts message
puts message
puts message
```

This program prints the string `Hello Penny!` three times to the standard output, each followed by a line break.

The name of a variable always starts with a lowercase [Unicode](https://en.wikipedia.org/wiki/Unicode) letter (or an underscore, but that's reserved for special use cases) and can otherwise consist of alphanumeric characters or underscores. As a typical convention, upper-case letter are avoided and names are written in [`snake_case`](https://en.wikipedia.org/wiki/Snake_case).

!!! note
    The kind of variables this lesson discusses are called *local variables*.
    Other kinds will be introduced later. For now we focus on local variables only.

## Type

> The type of a variable is automatically inferred by the compiler. In the above example, it's [`String`](https://crystal-lang.org/api/String.cr).

Reviewer: ```suggestion
The type of a variable is automatically inferred by the compiler. In the above example, it's [`String`](https://crystal-lang.org/api/latest/String.html).
```

You can verify this with [`typeof`](https://crystal-lang.org/api/toplevel.html#typeof(*expression):Class-class-method):

```{.crystal .crystal-play}
message = "Hello Penny!"

p! typeof(message)
```

!!! note

>     [`p!`](https://crystal-lang.org/api/toplevel.html#p!(*exps)-macro) is similar to `puts` as it prints the value to the standard output, but it also prints the expression in code. This makes it a useful tool for inspecting the state of a Crystal program and debugging.

Reviewer: ```suggestion
    [`p!`](https://crystal-lang.org/api/latest/toplevel.html#p!(*exps)-macro) is similar to `puts` as it prints the value to the standard output, but it also prints the expression in code. This makes it a useful tool for inspecting the state of a Crystal program and debugging.
```

## Reassigning a Value

A variable can be reassigned with a different value:

```{.crystal .crystal-play}
message = "Hello Penny!"

p! message

message = "Hello Sheldon!"

p! message
```

This also works with values of a different type. The type of the variable changes when a value of a different type is assigned. The compiler is smart enough to know which type it has at which point in the program.

```{.crystal .crystal-play}
message = "Hello Penny!"

> p! message, typeof(message)

Reviewer: (idea/optional) Maybe we should include the expected output as inline comments we could use later to verify the example are still valid.

Author: Yeah, I thought about that, too. On the other hand, if we show the expected output directly, it's less fun to click the button and actually see the code work.
I'd suggest to leave it for now and revisit this later.

message = 73

> p! message, typeof(message)

Reviewer: I would like to mention the difference between `typeof(message)` and `message.class` but to do that we need a loop at least to make a meaningful example, but that is not covered yet. So, something to keep in mind for the future.

# Variables

In order to store a value and re-use it later, it can be assigned to a variable.

For example, if you want to say `Hello Penny!` three times, you don't need to repeat the same string multiple times.
Instead, you can assign it to a local variable and re-use it:

```{.crystal .crystal-play}
message = "Hello Penny!"

puts message
puts message
puts message
```

This program prints the string `Hello Penny!` three times to the standard output, each followed by a line break.

> The name of a local variable always starts with a lowercase [Unicode](https://en.wikipedia.org/wiki/Unicode) letter (or an underscore, but that's reserved for special use cases) and can otherwise consist of alphanumeric characters or underscores. As a typical convention, upper-case letter are avoided and names are written in `snake_case`.

Reviewer 2: ```suggestion
The name of a local variable always starts with a lowercase [Unicode](https://en.wikipedia.org/wiki/Unicode) letter (or an underscore, but that's reserved for special use cases) and can otherwise consist of alphanumeric characters or underscores. As a typical convention, upper-case letter are avoided and names are written in [`snake_case`](https://en.wikipedia.org/wiki/Snake_case).
```
